doc/developer/design/20230110_window_functions.md

```diff
@@ -0,0 +1,592 @@
+
+The actual implementation of `aggregate` proceeds in similar steps, but it builds _A_ directly, rather than _A'_: rather than starting from all possible 1-length intervals, it starts from only those indexes that actually occur in the input data. This means that each merge step will find that it has to merge either 1 or 2 intervals, because one of the intervals that would have participated in the merge if we were building _A'_ rather than _A_ might actually be absent, due to not containing any such index that occurs in the input data. When "merging" only 1 interval rather than 2, the sum of the "merged" interval will be the same value as that 1 constituent interval.
+
+*3. (Index bit vector prefixes. This one is useful for certain performance considerations.)*
+This definition directly defines _A_ rather than _A'_. Let's consider the indexes that occur in the input data as bit vectors. For each of _len ∈ 0 ..= b-1_, let's define the set *D_len* to be the set of distinct prefixes of length _len_ of all the index bit vectors. In other words, for each _len_, we take the first _len_ bits of each of the indexes, and form *D_len* by deduplicating all these index prefixes. Let *D* be the union of all *D_len* sets.
+_A_ will have exactly one interval for each element of *D*. The actual mapping (i.e., what element of _A_ belongs to each element of *D_len*) is not so important, since we will rely on just the sizes of the *D_len* sets for certain performance considerations. (The actual mapping is as follows: For each *len*, for each *d ∈ D_len*, *A* includes the interval whose size is *b-len* and starts at _d * 2^(b-len)_.) See a performance discussion relying on this definition [here](#Performance-and-optimizations).
+
+##### The `broadcast` function
+
+The `broadcast` function takes the precomputed sums from `aggregate` and a set of "queries", which are the set of indexes for which we are interested in the prefix sum ending at that index.
+
+Just by looking at the bit patterns in the queries, it computes the set of intervals (`requests`) from which it will put together the final sums. Many precomputed intervals might participate in the final sum calculation for more than one query, and therefore `requests` are deduplicated.
+
+For a general prefix sum, it can happen that some intervals in `requests` don't occur among the pre-aggregated intervals, meaning that the sum for these intervals is 0. Therefore, there is an outer join to add these 0s. However, this case cannot occur when we use prefix sum for window functions, because our `queries` and the original set of indexes from which `aggregate` started are actually the same sets. Therefore, this outer join can be left out.
```

Comment:
This paragraph suggests a change to DD's implementation of `broadcast`. I am wondering here if we are at the end going to use a Materialize-rolled copy of DD's prefix sum implementation with several changes or if we are planning to reuse DD's implementation directly in all (some?) cases? Perhaps we could be a bit more explicit regarding implementation strategy, either in this or another section?

Reply:
We'll definitely need a Materialize-rolled copy. I added a sentence in the first paragraph of the `### Prefix Sum` section to make this clear.